### PR TITLE
Auto-initialize deques

### PR DESCRIPTION
**Pull request by zevv**

### CODE REVIEW

File: lib/pure/collections/deques.nim [hunk 1]
```diff
@@ -62,6 +62,12 @@ type
     data: seq[T]
     head, tail, count, mask: int
 
+proc initImpl[T](deq: var Deque[T], initialSize: int = 4) =
+  if deq.mask == 0:
+    assert isPowerOfTwo(initialSize)
```
Comment (juancarlospaco) [inline lib/pure/collections/deques.nim]: Should it be a `doAssert` :grey_question: :thinking:

Comment (zevv) [reply lib/pure/collections/deques.nim]: I think not. This was already there, I just moved the implementation to a separate proc.

**Comment by b3liever** [comment]
tablesimpl.nim does this differently:
First it has a template with the init code
```
template initImpl(result: typed, size: int)
```
and uses that directly in ``init`` procs. and then a second template that has the if check:
```nim
template checkIfInitialized() =
  when compiles(defaultInitialSize):
    if t.dataLen == 0:
      initImpl(t, defaultInitialSize)
```
which is used in ``[]``, ``[]=`` and etc.

**Comment by Araq** [comment]
Yeah, I also find the tableimpl.nim way more elegant.

File: lib/pure/collections/deques.nim [hunk 2]
```diff
@@ -121,6 +125,7 @@ proc `[]`*[T](deq: var Deque[T], i: Natural): var T {.inline.} =
     assert a[3] == 40
     doAssertRaises(IndexError, echo a[8])
 
+  initImpl(deq)
```
Comment (Araq) [inline lib/pure/collections/deques.nim]: Accessors shouldn't auto-init.

File: lib/pure/collections/deques.nim [hunk 3]
```diff
@@ -164,6 +170,7 @@ proc `[]`*[T](deq: var Deque[T], i: BackwardsIndex): var T {.inline.} =
     assert a[^4] == 20
     doAssertRaises(IndexError, echo a[^9])
 
+  initImpl(deq)
```
Comment (Araq) [inline lib/pure/collections/deques.nim]: Accessors shouldn't auto-init.

File: lib/pure/collections/deques.nim [hunk 4]
```diff
@@ -370,6 +379,7 @@ proc popFirst*[T](deq: var Deque[T]): T {.inline, discardable.} =
     assert a.popFirst == 10
     assert $a == "[20, 30, 40, 50]"
 
+  initImpl(deq)
```
Comment (Araq) [inline lib/pure/collections/deques.nim]: Accessors shouldn't auto-init.

File: lib/pure/collections/deques.nim [hunk 5]
```diff
@@ -395,6 +405,7 @@ proc popLast*[T](deq: var Deque[T]): T {.inline, discardable.} =
     assert a.popLast == 50
     assert $a == "[10, 20, 30, 40]"
 
+  initImpl(deq)
```
Comment (Araq) [inline lib/pure/collections/deques.nim]: Accessors shouldn't auto-init.

**Comment by Araq** [comment]
Ping @zevv

**Comment by zevv** [comment]
sorry, travels and family business getting in the way...

**Comment by zevv** [comment]
Removed init on accessors, refactored to use similar structure as `tablesImpl` (not sure why this is better, but it is at least more consistent)

**Pull request merged**